docs/ROADMAP.md

```diff
@@ -0,0 +1,382 @@
+# hyper 1.0 Roadmap
+
+## Goal
+
+Align current hyper to the [hyper VISION][VISION].
+
+The VISION outlines a decision-making framework, use-cases, and general shape
+of hyper. This roadmap describes the currently known problems with hyper, and
+then shows what changes are needed to make hyper 1.0 look more like what is in
+the VISION.
+
+## Known Issues
+
+
+> **Note**: These known issues are as of hyper v0.14.x. After v1.0 is released,
+ideally these issues will have been solved. Keeping this history may be helpful
+to Future Us, though.
+
+### Higher-level Client and Server problems
+
+Both the higher-level `Client` and `Server` types have stability concerns.
+
+For the `hyper::Server`:
+
+- The `Accept` trait is complex, and too easy to get wrong. If used with TLS, a
+slow TLS handshake can affect all other new connections waiting for it to
+finish.  - The `MakeService<&IO>` is confusing. The bounds are an assault on
+the eyes.  - The `MakeService` API doesn't allow to easily annotate the HTTP
+connection with `tracing`.  - Graceful shutdown doesn't give enough control.
+
+
+It's more common for people to simply use `hyper::server::conn` at this point,
+than to bother with the `hyper::Server`.
+
+While the `hyper::Client` is much easier to use, problems still exist:
+
+- The whole `Connect` design isn't stable.
+  - ALPN and proxies can provide surprising extra configuration of connections.
+  - Some `Connect` implementations may wish to view the path, in addition to    the scheme, host, and port.
+  - Wants `runtime` feature
+- The Pool could be made more general or composable. At the same time, more customization is
+  desired, and it's not clear
+how to expose it yet.
+
+
+### Runtime woes
+
+hyper has been able to support different runtimes, but it has sometimes awkward
+default support for Tokio.
+
+- The `runtime` cargo-feature isn't additive
+- Built-in Tokio support can be confusing
```

Comment:
as someone who [thinks it's fun to build async runtimes](https://git.asonix.dog/safe-async) i selfishly want hyper to keep working with generic runtime-ish traits

Reply:
I also think it's a good goal to enable hyper to be used on other runtimes so GTK (and other desktop frameworks) can make use of hyper on their one runtime, e.g. https://docs.rs/glib/latest/glib/struct.MainContext.html#method.spawn

Reply:
again, it is not possible to do without network primitives being portable across runtimes.

Reply:
maybe not for h3, but h1 at least already works in hyper on 3rd party runtimes.

Reply:
hyper uses tokio's primitives for that, tokio is required dependency there.

Reply:
Ah, I misunderstood "primitive" to mean types like TcpStream, but you were referring to AsyncRead and AsyncWrite. I agree that this is an acceptable compromise.

Reply:
Actually I also meant TcpStream and friends as they are tied to runtime

Reply:
Tokio's TcpStream is not currently required to use hyper. I've implemented an adapter to run hyper clients and servers on my own custom runtime in 0.14 and I did not need tokio's TcpStream for that.

https://git.asonix.dog/safe-async/hyperjive/

Reply:
Yes, hyper v0.14 already works with other runtimes, since you can provide your own IO. I've seen it adapted to Google's Fuchsia, and the C API enables curl to adapt its own runtime.

Reply:
note that Fuchsia currently uses hyper v.0.13, but we're in the process of trying to update it to 0.14 in https://fuchsia-review.googlesource.com/c/fuchsia/+/611683.